### PR TITLE
refactor: Fix unreachable code in init arg checks

### LINKED ISSUE #19017: Fix -Wunreachable-code-loop-increment warning
In #19015 it's been suggested that we add some new compiler warnings to our build. Some of these, such as `-Wunreachable-code-loop-increment`, generate warnings. We'll likely want to fix these up if we're going to turn these warnings on.

```bash
init.cpp:969:5: warning: loop will run at most once (loop increment never executed) [-Wunreachable-code-loop-increment]
    for (const auto& arg : gArgs.GetUnsuitableSectionOnlyArgs()) {
    ^~~
1 warning generated.
```

https://github.com/bitcoin/bitcoin/blob/aa8d76806c74a51ec66e5004394fe9ea8ff0fac4/src/init.cpp#L968-L972

If you want to build master, with `-Wunreachable-code-loop-increment`, you can use something like this:
```bash
./autogen.sh
./configure CXXFLAGS="-Wunreachable-code-loop-increment"
make check
```

### PR DESCRIPTION
**Pull request by jonathanschoeller**
Closes: #19017 

In #19015 it's been suggested that we add some new compiler warnings to our build. Some of these, such as `-Wunreachable-code-loop-increment`, generate warnings. We'll likely want to fix these up if we're going to turn these warnings on.

```shell
init.cpp:969:5: warning: loop will run at most once (loop increment never executed) [-Wunreachable-code-loop-increment]
     for (const auto& arg : gArgs.GetUnsuitableSectionOnlyArgs()) {
     ^~~
 1 warning generated.
 ```
 https://github.com/bitcoin/bitcoin/blob/aa8d76806c74a51ec66e5004394fe9ea8ff0fac4/src/init.cpp#L968-L972

To fix this, collect all errors, and output them in a single error message after the loop completes. This resolves the unreachable code warning, and avoids popup hell that could result from outputting a seperate message for each error or warning one by one.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: thanks for fixing this and thereby allowing us to catch more severe issues using `-Wunreachable-code-loop-increment` going forward!

Great first time contribution and great rationale section in the PR description.

Warm welcome as a contributor @jonathanschoeller! :)

File: src/init.cpp [hunk 1]
```diff
@@ -964,17 +964,29 @@ bool AppInitParameterInteraction()
 
     // also see: InitParameterInteraction()
 
-    // Warn if network-specific options (-addnode, -connect, etc) are
+    // Error if network-specific options (-addnode, -connect, etc) are
     // specified in default section of config file, but not overridden
     // on the command line or in this network's section of the config file.
     std::string network = gArgs.GetChainName();
-    for (const auto& arg : gArgs.GetUnsuitableSectionOnlyArgs()) {
```
Comment (MarcoFalke) [inline src/init.cpp:971]: I don't think this line needs to change. Below you can simply check for `errors.empty()`, which should do the exactly same.

File: src/init.cpp [hunk 2]
```diff
@@ -964,17 +964,29 @@ bool AppInitParameterInteraction()
...
+    for (const auto& arg : unsuitable_args) {
+        errors += strprintf(_("Config setting for %s only applied on %s network when in [%s] section.") + Untranslated("\n"), arg, network, network);
+    }
+
+    if (!unsuitable_args.empty()) {
+        return InitError(errors);
     }
 
     // Warn if unrecognized section name are present in the config file.
-    for (const auto& section : gArgs.GetUnrecognizedSections()) {
```
Comment (MarcoFalke) [inline src/init.cpp:976]: Same

**Review by maflcko — APPROVED** [review]
Nice. ACK with or without the nit addressed

**Comment by MarcoFalke** [comment]
Oh, it looks like you might have to adjust the tests for the longer error messages.

File: src/init.cpp (hunk 1 above)
Comment (jonathanschoeller) [reply src/init.cpp:971]: `class "bilingual_str" has no member "empty"`. 

Would it be reasonable to use `errors.translated.empty()` for this purpose?

Comment (hebasto) [reply src/init.cpp:971]: > `class "bilingual_str" has no member "empty"`.
> 
> Would it be reasonable to use `errors.translated.empty()` for this purpose?

Feel free to get it from https://github.com/bitcoin/bitcoin/pull/18927/commits/1df513531f04ef31d55d2fa2acc146d526745788 :)

**Comment by hebasto** [comment]
Concept ACK.

File: src/init.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/init.cpp:971]: Oh, is that not yet merged. Sorry didn't realize

File: src/init.cpp [hunk 3]
```diff
@@ -964,17 +964,27 @@ bool AppInitParameterInteraction()
...
+
+    if (!errors.empty()) {
+        return InitError(errors);
     }
 
     // Warn if unrecognized section name are present in the config file.
+    bilingual_str warnings;
     for (const auto& section : gArgs.GetUnrecognizedSections()) {
-        InitWarning(strprintf(Untranslated("%s:%i ") + _("Section [%s] is not recognized."), section.m_file, section.m_line, section.m_name));
+        warnings += strprintf(Untranslated("%s:%i ") + _("Section [%s] is not recognized.\n"), section.m_file, section.m_line, section.m_name);
```
Comment (MarcoFalke) [inline src/init.cpp:983]: please don't change the translation string `_("Section [%s] is not recognized.")` and keep the special character `\n` untranslated (like above)

Comment (hebasto) [inline src/init.cpp:983]:
```suggestion
        warnings += strprintf(Untranslated("%s:%i ") + _("Section [%s] is not recognized.") + Untranslated("\n"), section.m_file, section.m_line, section.m_name);
```

**Review by hebasto — CHANGES_REQUESTED** [review]
ACK b546870d178222acd5d3f131410b8846f4875412, tested on Linux Mint 19.3 (x86_64).

I suggest to add a commit to modify `configure.ac` as well:
```diff
--- a/configure.ac
+++ b/configure.ac
@@ -391,6 +391,7 @@ if test "x$enable_werror" = "xyes"; then
   dnl https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78010
   AX_CHECK_COMPILE_FLAG([-Werror=suggest-override],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=suggest-override"],,[[$CXXFLAG_WERROR]],
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
+  AX_CHECK_COMPILE_FLAG([-Werror=unreachable-code-loop-increment],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=unreachable-code-loop-increment"],,[[$CXXFLAG_WERROR]])
 fi
 
 if test "x$CXXFLAGS_overridden" = "xno"; then
@@ -410,6 +411,7 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wsign-compare],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsign-compare"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wsuggest-override],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wsuggest-override"],,[[$CXXFLAG_WERROR]],
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
+  AX_CHECK_COMPILE_FLAG([-Wunreachable-code-loop-increment],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunreachable-code-loop-increment"],,[[$CXXFLAG_WERROR]])
 
   dnl Some compilers (gcc) ignore unknown -Wno-* options, but warn about all
   dnl unknown options if any other warning is produced. Test the -Wfoo case, and
```

File: src/init.cpp (hunk 3 above)
Comment (hebasto) [reply src/init.cpp:983]: Slow [typing](https://github.com/bitcoin/bitcoin/pull/19131#pullrequestreview-421772387) from my side :)

**Comment by jonathanschoeller** [comment]
Thanks @practicalswift , @MarcoFalke , and @hebasto for welcoming me and taking the time to help me with this PR.

**Review by hebasto — APPROVED** [review]
re-ACK eea81146571480b2acd12c8cd7f36b04d056c47f, only suggested changes applied since the [previous](https://github.com/bitcoin/bitcoin/pull/19131#pullrequestreview-421772387) review.

**Comment by laanwj** [comment]
Code review ACK eea81146571480b2acd12c8cd7f36b04d056c47f

**Pull request merged**